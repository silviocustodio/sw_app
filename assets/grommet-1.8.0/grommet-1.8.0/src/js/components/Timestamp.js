// (C) Copyright 2014-2017 Hewlett Packard Enterprise Development LP

import React, { Component } from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import CSSClassnames from '../utils/CSSClassnames';
import { getCurrentLocale } from '../utils/Locale';

const CLASS_ROOT = CSSClassnames.TIMESTAMP;

function _showField(field, fields) {
  let result = true;
  if (fields) {
    if (Array.isArray(fields)) {
      result = fields.indexOf(field) !== -1;
    } else {
      result = field === fields;
    }
  }
  return result;
}

export default class Timestamp extends Component {

  constructor(props, context) {
    super(props, context);
    this._formatForLocale = this._formatForLocale.bind(this);
    this.state = {};
  }

  componentDidMount () {
    this._formatForLocale(this.props);
  }

  componentWillReceiveProps (nextProps) {
    this._formatForLocale(nextProps);
  }

  _formatForLocale ({value, fields}) {
    const locale = getCurrentLocale();
    const dateObj = moment(value).locale(locale);

    let dateFormat;
    let yearFormat;
    let monthFormat;
    let dayFormat;

    let timeFormat;
    let hourFormat;
    let minuteFormat;
    let secondFormat;

    if (_showField('date', fields)) {
      dateFormat = 'll';
    }

    if (!dateFormat) {
      if (_showField('year', fields)) {
        yearFormat = 'YYYY';
      }

      if (_showField('month', fields)) {
        monthFormat = 'MMM';
      } else if (_showField('month-full', fields)) {
        monthFormat = 'MMMM';
      }

      if (_showField('day', fields)) {
        dayFormat = 'D';
      }
    } else if (_showField('month-full', fields)) {
      dateFormat = 'LL';
    }

    if (_showField('time', fields)) {
      timeFormat = 'LT';
    } 

    if (!timeFormat) {
      if (_showField('hour', fields) || _showField('hours', fields)) {
        hourFormat = 'hh';
      }
      if (_showField('minute', fields) || _showField('minutes', fields)) {
        minuteFormat = (hourFormat ? ':' : '') + 'mm';
      }
      if (_showField('second', fields) || _showField('seconds', fields)) {
        secondFormat = (minuteFormat ? ':' : '') + 'ss';
      }
    } else if (_showField('second', fields) || _showField('seconds', fields)) {
      timeFormat = 'LTS';
    }

    if (!dateFormat) {
      dateFormat = (
        `${monthFormat || ''} ${dayFormat || ''} ${yearFormat || ''}`
      );
    }

    if (!timeFormat) {
      timeFormat = (
        `${hourFormat || ''}${minuteFormat || ''}${secondFormat || ''}`
      );
    }

    const date = dateFormat !== '  ' ? dateObj.format(dateFormat) : undefined;
    const time = timeFormat !== '' ? dateObj.format(timeFormat) : undefined;

    this.setState({ date, time });
  }

  render () {
    const { align, className, ...props } = this.props;
    const { date, time } = this.state;
    delete props.fields;
    delete props.value;
    const classes = classnames(
      CLASS_ROOT,
      {
        [`${CLASS_ROOT}--${align}`]: align
      },
      className
    );

    let dateElement;
    if (date) {
      dateElement = <span className={`${CLASS_ROOT}__date`}>{date}</span>;
    }

    let timeElement;
    if (time) {
      timeElement = <span className={`${CLASS_ROOT}__time`}>{time}</span>;
    }

    return (
      <span {...props} className={classes}>
        {dateElement}
        {timeElement}
      </span>
    );
  }

}

const FIELD_TYPES = PropTypes.oneOf([
  'date', 'time', 'year', 'month', 'month-full', 'day',
  'hour', 'minute', 'second',
  'hours', 'minutes', 'seconds' // deprecated
]);

Timestamp.propTypes = {
  align: PropTypes.oneOf(['start', 'center', 'end']),
  fields: PropTypes.oneOfType([
    PropTypes.arrayOf(FIELD_TYPES),
    FIELD_TYPES
  ]),
  value: PropTypes.oneOfType([
    PropTypes.string, // ISO-8601 string
    PropTypes.object  // Date object
  ]).isRequired
};

Timestamp.defaultProps = {
  fields: ["date", "time"]
};
