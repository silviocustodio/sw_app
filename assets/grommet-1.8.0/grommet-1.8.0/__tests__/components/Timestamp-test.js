// (C) Copyright 2014-2016 Hewlett Packard Enterprise Development LP

import React from 'react';
import renderer from 'react-test-renderer';

import Timestamp from '../../src/js/components/Timestamp';

describe('Timestamp', () => {
  it('has correct default options', () => {
    const component = renderer.create(
      <Timestamp value='4/7/2015 10:00:01 am' />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('displays time', () => {
    const component = renderer.create(
      <Timestamp value='4/7/2015 10:00:01 am' fields={"time"} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('displays time with seconds', () => {
    const component = renderer.create(
      <Timestamp value='4/7/2015 10:00:01 am' fields={["time", "seconds"]} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('displays date', () => {
    const component = renderer.create(
      <Timestamp value='4/7/2015 10:00:01 am' fields={"date"} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('displays date and time', () => {
    const component = renderer.create(
      <Timestamp value='4/7/2015 10:00:01 am' fields={["date", "time"]} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('displays date, time, seconds', () => {
    const component = renderer.create(
      <Timestamp value='4/7/2015 10:00:01 am' fields={
        ["date", "time", "seconds"]
      } />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('displays a classname', () => {
    const component = renderer.create(
      <Timestamp value='4/7/2015 10:00:01 am' className="test" />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('aligns to end', () => {
    const component = renderer.create(
      <Timestamp value='4/7/2015 10:00:01 am' align="end" />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
