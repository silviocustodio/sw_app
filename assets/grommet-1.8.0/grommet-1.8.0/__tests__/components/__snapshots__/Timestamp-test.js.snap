// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timestamp aligns to end 1`] = `
<span
  className="grommetux-timestamp grommetux-timestamp--end"
>
  <span
    className="grommetux-timestamp__date"
  >
    Apr 7, 2015
  </span>
  <span
    className="grommetux-timestamp__time"
  >
    10:00 AM
  </span>
</span>
`;

exports[`Timestamp displays a classname 1`] = `
<span
  className="grommetux-timestamp test"
>
  <span
    className="grommetux-timestamp__date"
  >
    Apr 7, 2015
  </span>
  <span
    className="grommetux-timestamp__time"
  >
    10:00 AM
  </span>
</span>
`;

exports[`Timestamp displays date 1`] = `
<span
  className="grommetux-timestamp"
>
  <span
    className="grommetux-timestamp__date"
  >
    Apr 7, 2015
  </span>
</span>
`;

exports[`Timestamp displays date and time 1`] = `
<span
  className="grommetux-timestamp"
>
  <span
    className="grommetux-timestamp__date"
  >
    Apr 7, 2015
  </span>
  <span
    className="grommetux-timestamp__time"
  >
    10:00 AM
  </span>
</span>
`;

exports[`Timestamp displays date, time, seconds 1`] = `
<span
  className="grommetux-timestamp"
>
  <span
    className="grommetux-timestamp__date"
  >
    Apr 7, 2015
  </span>
  <span
    className="grommetux-timestamp__time"
  >
    10:00:01 AM
  </span>
</span>
`;

exports[`Timestamp displays time 1`] = `
<span
  className="grommetux-timestamp"
>
  <span
    className="grommetux-timestamp__time"
  >
    10:00 AM
  </span>
</span>
`;

exports[`Timestamp displays time with seconds 1`] = `
<span
  className="grommetux-timestamp"
>
  <span
    className="grommetux-timestamp__time"
  >
    10:00:01 AM
  </span>
</span>
`;

exports[`Timestamp has correct default options 1`] = `
<span
  className="grommetux-timestamp"
>
  <span
    className="grommetux-timestamp__date"
  >
    Apr 7, 2015
  </span>
  <span
    className="grommetux-timestamp__time"
  >
    10:00 AM
  </span>
</span>
`;
